### ISSUE TITLE
Offline testing without external resources

### ISSUE THREAD

**Issue by peterdd** [pongstylin/tactics#7]
Got it running in vagrant box with opensuse15 under 10.0.0.x:2000 , but requires accessing googleapis.com , so not playable/testable when no internet connection.

**Comment by pongstylin** [comment]
While the dependency on google connectivity can be lifted, the original TAO graphics are deliberately excluded from the repository, so I see no development path for removing all internet dependence. At least, not until original artwork can be created.

**Comment by peterdd** [comment]
Local cache/storage of artwork? An own simple (abstract like simple svg shapes) default theme with included free artwork and the original artwork optional loaded and stored?

**Comment by pongstylin** [comment]
Hm, well, I don't think the project is mature enough for such an abstraction right now.  But, if it is difficult to work on the project, I'll be happy to explore that difficulty and an alternative solution.

**Comment by Enijar** [comment]
We could cache the images on initial load inside IndexedDB. I've written a quick example of what this could look like [here](https://jsbin.com/judopecaya/edit?js,console). Basically the idea is to load all assets as blobs and cache them inside IndexedDB on the user's browser with a key. This is better for local development and user experience, since assets only need to be loaded once; on subsequent requests the assets will load almost instantly. This also saves network traffic for subsequent loads.

**Comment by pongstylin** [comment]
Right now, I believe the assets have a 1-day cache.  So, multiple loads on the same day should not require extra network traffic.  This cache period can be increased, as well.  There is no need to implement a different storage mechanism simply to reduce traffic.  The old fashioned cache mechanism was created for exactly that purpose.

With that said, there is a feature that would be useful: "Offline Mode".  Using "Service Workers" and the "Cache API" we could implement local storage of assets.  One caveat is that "Service Workers" requires SSL or a localhost URL to function.  So, to publicly release these changes I would have to offer an HTTPS-enabled site, which is fine.

**Comment by Enijar** [comment]
The 1-day cache on images is only useful for users. That's not useful for devs who have their cache disabled by default and have to download the images on each page load. This caching system I made is similar to the saw service workers work but we have more control over the cache this way. We can force update the cache for users by simply increasing the version number of the IndexedDB cache.

**Comment by Enijar** [comment]
On second thoughts we should make a script to download all images for local development. This was the load time for devs when the cache is disabled is much faster since outbound network requests are not required. 

Thoughts?

**Comment by pongstylin** [comment]
Yeah, I understand the part about the cache being disabled during development.  If the load times really suck, then I suppose it is worthwhile.  Right now, each of the two apps only load the resources that it needs (the Faceoff app does not load egg and dragon images).  I prefer that it remains "on-demand" like that as opposed to a "load everything" approach (for now, that might change).  I'm willing to consider an IndexedDB solution for now even if it is ultimately replaced with an "Offline Mode" for users.  I do not know if IndexedDB can be the be-all solution for all use cases.

**Comment by Enijar** [comment]
We don't have to go with an IndexedDB approach just now, since our user-base is only developers. We can go with the build script to download images and serve them locally for development mode.

**Comment by latin-programmer** [comment]
![fixme](https://user-images.githubusercontent.com/3208425/47830620-af247f80-dd5a-11e8-9c9b-d73288f19b46.png)

I add my vote on the idea of "An own simple (abstract like simple svg shapes) default theme with included free artwork and the original artwork optional loaded and stored". 

An idea: 
 - Develop a "fallback" plan for the URL loading. If it doesn't exist in taorankings, try the same location on the static directory using the same path
 - Accept pull requests for images in said static directory. Those would be for dev purposes only and would not be the actual proprietary images (better that my "fix me" art though ;P). You can validate this before accepting the pull request and rejecting it if needed. 
 - Those images would be "promises" for the final images we are working on. Once you publish them in taorankings, all of us would start seeing the actual images on the build. 

Just a thought, but I did want to state that contributing on the images is a bit more involved that what it could be if there was something in the code that has out-of-the box for placeholders for missing images. Right now for the screenshot above, I had to do some tinkering and I could make it work from there for a simple unit, but for the ones that require more images, it would be a more painful process.

**Comment by Enijar** [comment]
Due to the large number of image sprites, I don't think they should be committed to version control. I think the best approach for images would be:

- Host images online (currently http://www.taorankings.com/html5/units/x/color/x.png)
- Have a build script for code contributors to run and download the files to their machine
- In the config, have a check to see if the current NODE_ENV is dev or prod
- NODE_ENV dev will load the downloaded images and prod will load the images from the network

The build script could look something like this (with Makefile):

```Makefile
PROJECT = "Tactics"

install:
	@echo "Installing ${PROJECT}..." \
	npm install

start:
	@echo "Starting ${PROJECT}..." \
	@NODE_ENV=development \
	npm start

dev:
	@echo "Starting ${PROJECT}..." \
	@NODE_ENV=development \
	npm run dev

.PHONY: install start dev
```

**Comment by pongstylin** [comment]
Ultimately, sprites will be combined into larger image files (spritesheets) to improve download times and asset organization. The image files will be accompanied by data files that describe how to position and combine sprites so that a complete unit can be rendered on a tile, or even how it can be animated. This will require less tinkering with code and easier management of assets. This is what I was referring to earlier when I said this project is not very mature yet, which made it difficult for me to think about abstracting data access.

**Comment by latin-programmer** [comment]
Another idea I floated around if you are concerned about committing a bunch of images to the repo: 
 - Don't have multiple placeholder images, have a single blank placeholder image
 - Write fall back code that does the following if an image doesn't exist: 
 -- Log the image name in the console
 -- Load the placeholder image instead 
 -- Place text in the place holder image indicating some hints of what is needed (what unit, what color, at the least, some more details if possible). 

If you had such an implementation in place, you can pretty much implement the brains of the game just by using place holders. Of course you would have a white box that say "Dark Magic Witch - White" that is moving around. But would you mind if you are all the sudden getting ahead on the implementation while some of us work on getting the images ready (and in turn are able to see how they are going to be used).

**Comment by Enijar** [comment]
Hmm, that could be one way to progress the development of the core-game logic. I was thinking more along the lines of what should and shouldn't be committed to version control. From looking at what has already been done and what needs to be done it seems like the most time consuming part of this project will be extracting the animation sprite for all units.

**Comment by pongstylin** [comment]
If you want to create a bunch of units that use placeholder images, that is fine.  I suggest creating a 3rd app separate from the current Faceoff and Chaos apps.  As much as possible, keep unit logic separate from rendering logic with the goal that your unit logic will still be useful if we were to swap in different images/animations.  Unit logic includes:
# Initial and modified stats such as health, blocking, defense
# Board placement such as current tile and the direction the unit is facing.
# Unit capabilities such movement range, attack vectors, and special attacks.
# Triggered effects, such as reduced health or status effects on other units.

As Enijar suggests, unit logic isn't actually very hard or time consuming.  But, even if the images are placeholders and the unit "teleports" from A to B when you move it and status effects are applied instantly upon making an attack, I think people would still enjoy playing with it - especially if they can play each other.  If you want to start working on player vs player, then create a new issue and we can start working out a rough framework for it.

**Comment by pongstylin** [comment]
I'm closing out this issue.  While the original request seemed to suggest the removal of all external resources, the conversation had developed to the point where we just don't like how long it takes to load all the resources while loading the game with development tools open (which disables cache).

The latter issue is resolved.  While adding "Add to Home Screen" prompt support for Chrome, I also implemented persistent caching of resources.  But, while developing, it only caches external resources (not the local ones, which change all the time).  This means the game loads a lot more quickly since the resources are cached DESPITE disabling the cache.
